recipes/openai.py

```diff
@@ -0,0 +1,354 @@
+    # Even if the model is None, this call is used as a healthcheck to verify access.
+    headers = {
+        "Authorization": f"Bearer {api_key}",
+        "OpenAI-Organization": org,
+    }
+    r = retry(
+        lambda: httpx.get(
+            "https://api.openai.com/v1/models",
+            headers=headers,
+        ),
+        n=1,
+        timeout_s=1,
+    )
+    if r.status_code == 422:
+        m = (
+            "Could not access api.openai.com -- 422 permission denied."
+            "Visit https://beta.openai.com/account/api-keys to check your API keys."
+        )
+        msg.fail(m)
+        sys.exit(-1)
+    elif r.status_code != 200:
+        m = "Error accessing api.openai.com" f"{r.status_code}: {r.text}"
+        msg.fail(m)
+        sys.exit(-1)
+
+    if model is not None:
+        response = r.json()["data"]
+        models = [response[i]["id"] for i in range(len(response))]
+        if model not in models:
+            e = f"The specified model '{model}' is not available. Choices are: {sorted(set(models))}"
+            msg.fail(e, exits=1)
+
+    return api_key, org
+
+
+def read_prompt_examples(
+    path: Optional[Path], *, example_class: PromptExample
+) -> List[PromptExample]:
+    if path is None:
+        return []
+    elif path.suffix in (".yml", ".yaml"):
+        return read_yaml_examples(path, example_class=example_class)
+    elif path.suffix == ".json":
+        data = srsly.read_json(path)
+        assert isinstance(data, list)
+        return [PromptExample(**eg) for eg in data]
+    else:
+        msg.fail(
+            "The --examples-path (-e) parameter expects a .yml, .yaml or .json file."
+        )
+        sys.exit(-1)
+
+
+def load_template(path: Path) -> jinja2.Template:
+    # I know jinja has a lot of complex file loading stuff,
+    # but we're not using the inheritance etc that makes
+    # that stuff worthwhile.
+    if not path.suffix == ".jinja2":
+        msg.fail(
+            "The --prompt-path (-p) parameter expects a .jinja2 file.",
+            exits=1,
+        )
+    with path.open("r", encoding="utf8") as file_:
+        text = file_.read()
+    return jinja2.Template(text, undefined=jinja2.DebugUndefined)
+
+
+def retry(
+    call_api: Callable[[], httpx.Response],
+    n: int,
+    timeout_s: int,
+    error_codes: List[int] = [429, 503],
+) -> httpx.Response:
+    """Retry a call to the OpenAI API if we get a non-ok status code.
+
+    This function automatically retries a request if it catches a response
+    with an error code in `error_codes`. The amount of timeout also increases
+    exponentially every time we retry.
+    """
+    assert n >= 0
+    assert timeout_s >= 1
+    r = call_api()
+    i = -1
+    # We don't want to retry on every non-ok status code. Some are about
+    # incorrect inputs, etc. and we want to terminate on those.
+    while i < n and r.status_code in error_codes:
```

Comment:
> I noticed httpx.ReadTimeout errors as well. They seemed to break the flow just around 60 requests.

My gut fix would be to increase the timeout. But yeah, I'll figure out why it happens with just 60 requests.

Reply:
I'm finding these errors to be pretty random. They can appear all of the sudden and go away quickly.